index.js

```diff
@@ -104,14 +103,45 @@ var postKinesisStream = function(records, exportFile, schemaData){
     Records: records_in_avro_format, /* required */
     StreamName: exportFile.postStream /* required */
   }
-  /*console.log("Not posting records")
-  console.log('Records to kinesis - ' + records_in_avro_format.length)*/
-  kinesis.putRecords(params, function (err, data) {
-    if (err) console.log(err, err.stack); // an error occurred
-    else    {
-      postedToKinesis += records_in_avro_format.length;
-      console.log("Successfully posted to kinesis: " + postedToKinesis);           // successful response
-    }
-  })
 
+  if(putRecordsUntilCompletelySuccessful (params, exportFile, true, records_in_avro_format)){
+    postedToKinesis += records_in_avro_format.length;
+    console.log("Successfully posted to kinesis: " + postedToKinesis); 
+  }
 }
+
+  var putRecordsUntilCompletelySuccessful = (params, exportFile, startRunning, recordsToPost) => {
+    if(startRunning) {
+      console.log("Records to post count: " + recordsToPost.length)
+      kinesis.putRecords(params, (err, data) => {
+              if (err) {
+                console.log(err, err.stack); // an error occurred
+                return false
+              }
+              else{
+                if(data.FailedRecordCount > 0){
+                  console.log('Hit putRecords error. Going to retrieve failed records')
+                  var failedRecords = getFailedRecords(data, recordsToPost)
+                  var params = {
+                      Records: failedRecords,
+                      StreamName: exportFile.postStream
+                  }
+                  putRecordsUntilCompletelySuccessful(params, exportFile, true, failedRecords)
+                }else  startRunning = false 
```

Comment:
For some reason introducing the recursive functionality made me think of it as a while loop. Thanks for correcting this.